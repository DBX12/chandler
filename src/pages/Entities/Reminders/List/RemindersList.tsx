import moment from 'moment';
import React, {PureComponent} from 'react';
import {ActivityIndicator, FlatList, Text, View} from 'react-native';

import {EmptyActivity, Navbar} from '@components';
import {I18n} from '@i18n';
import {IReminder} from '@models';
import {IPopAction} from '@navigator/NavigationService';
import {commonStyles} from '@theme';
import {styles} from './RemindersList.styles';

interface IRemindersListProps {
  pop: IPopAction;
  getRemindersByContact: () => void;
  reminders: IReminder[];
  isFetching: boolean;
}

export class RemindersList extends PureComponent<IRemindersListProps, {}> {
  public componentWillMount() {
    this.props.getRemindersByContact();
  }

  public keyExtractor = (item, index) => String(item.id);

  public renderFooter = () => {
    const {isFetching} = this.props;

    if (!isFetching) {
      return null;
    }

    return (
      <ActivityIndicator style={commonStyles.activityIndicator} size="large" />
    );
  };

  public renderItem = ({item, index}) => {
    const {reminders} = this.props;
    const reminder = reminders[index];

    const nextDate = moment(reminder.next_expected_date);

    return (
      <View style={styles.reminderContainer}>
        <View style={styles.dateContainer}>
          <Text style={styles.date}>{nextDate.format('MMM DD')}</Text>
          <Text style={styles.dateYear}>{nextDate.format('YYYY')}</Text>
        </View>

        <View style={commonStyles.flex}>
          <Text>{reminder.title}</Text>
          {reminder.description ? (
            <Text style={styles.descriptionText}>{reminder.description}</Text>
          ) : null}

          <View style={commonStyles.row}>
            <Text style={styles.descriptionText}>{nextDate.fromNow()}</Text>
            <View style={styles.badge}>
              <Text style={styles.badgeText}>
                {I18n.t('reminders:frequency.' + reminder.frequency_type)}
              </Text>
            </View>
          </View>
        </View>
      </View>
    );
  };

  public render() {
    const {pop, reminders, getRemindersByContact, isFetching} = this.props;

    return (
      <View style={commonStyles.flex}>
        <Navbar title={I18n.t('reminders:reminders')} onBack={pop} />
        {isFetching || reminders.length ? (
          <FlatList
            data={reminders}
            renderItem={this.renderItem}
            keyExtractor={this.keyExtractor}
            ListFooterComponent={this.renderFooter}
            onEndReached={getRemindersByContact}
            onEndReachedThreshold={0.5}
          />
        ) : (
          <EmptyActivity
            image={require('./assets/empty-reminders.png')}
            title={I18n.t('reminders:emptyTitle')}
            subtitle={I18n.t('reminders:emptySubtitle')}
          />
        )}
      </View>
    );
  }
}
