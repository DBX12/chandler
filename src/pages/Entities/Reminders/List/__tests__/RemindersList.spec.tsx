import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import {RemindersList} from '../RemindersList';

describe('Pages', () => {
  describe('Reminders', () => {
    let pop;
    let getRemindersByContact;

    beforeEach(() => {
      pop = jest.fn();
      getRemindersByContact = jest.fn();
    });

    it('should renders correctly without reminders and fetching', () => {
      const reminders = [];
      const tree = shallow(
        <RemindersList
          pop={pop}
          getRemindersByContact={getRemindersByContact}
          isFetching={true}
          reminders={reminders}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
    });

    it('should renders correctly without reminders and not fetching', () => {
      const reminders = [];
      const tree = shallow(
        <RemindersList
          pop={pop}
          getRemindersByContact={getRemindersByContact}
          isFetching={false}
          reminders={reminders}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
    });

    it('should renders correctly with reminders and fetching', () => {
      const reminders = [
        {body: 'My body', created_at: '1991-12-17', frequency_type: 'year'},
      ];
      const tree = shallow(
        <RemindersList
          pop={pop}
          getRemindersByContact={getRemindersByContact}
          isFetching={true}
          reminders={reminders as any}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 0})))
      ).toMatchSnapshot();
    });

    it('should renders correctly with reminders and fetching', () => {
      const reminders = [
        {body: 'My body', created_at: '1991-12-17', frequency_type: 'month'},
        {
          body: 'My body 2',
          created_at: '1993-12-17',
          frequency_type: 'all_time',
        },
      ];
      const tree = shallow(
        <RemindersList
          pop={pop}
          getRemindersByContact={getRemindersByContact}
          isFetching={false}
          reminders={reminders as any}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 0})))
      ).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 1})))
      ).toMatchSnapshot();
    });

    it('should display the reminder indicator in the footer if it is fetching', () => {
      const reminders = [
        {body: 'My body', created_at: '1991-12-17'},
        {body: 'My body 2', created_at: '1993-12-17'},
      ];
      const tree = shallow(
        <RemindersList
          pop={pop}
          getRemindersByContact={getRemindersByContact}
          isFetching={true}
          reminders={reminders as any}
        />
      );
      expect(toJson(shallow(tree.instance().renderFooter()))).toMatchSnapshot();
    });

    it('should not display the reminder indicator in the footer if it is not fetching', () => {
      const reminders = [
        {body: 'My body', created_at: '1991-12-17'},
        {body: 'My body 2', created_at: '1993-12-17'},
      ];
      const tree = shallow(
        <RemindersList
          pop={pop}
          getRemindersByContact={getRemindersByContact}
          isFetching={false}
          reminders={reminders as any}
        />
      );
      expect(tree.instance().renderFooter()).toBeNull();
    });

    it('should have a keyExtractor', () => {
      const reminders = [{id: 1}, {id: 2}];
      const tree = shallow(
        <RemindersList
          pop={pop}
          getRemindersByContact={getRemindersByContact}
          isFetching={false}
          reminders={reminders as any}
        />
      );
      expect(tree.instance().keyExtractor(reminders[0])).toBe('1');
      expect(tree.instance().keyExtractor(reminders[1])).toBe('2');
    });
  });
});
