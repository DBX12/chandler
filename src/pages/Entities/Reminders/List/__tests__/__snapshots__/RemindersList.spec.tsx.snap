// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages Reminders should display the reminder indicator in the footer if it is fetching 1`] = `
<ActivityIndicator
  animating={true}
  color="#999999"
  hidesWhenStopped={true}
  size="large"
  style={
    Object {
      "marginVertical": 20,
    }
  }
/>
`;

exports[`Pages Reminders should renders correctly with reminders and fetching 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Reminders"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={
      Array [
        Object {
          "body": "My body",
          "created_at": "1991-12-17",
          "frequency_type": "year",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Component>
`;

exports[`Pages Reminders should renders correctly with reminders and fetching 2`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "flexDirection": "row",
      "paddingHorizontal": 15,
      "paddingVertical": 20,
    }
  }
>
  <Component
    style={
      Object {
        "paddingRight": 20,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#355974",
          "fontSize": 17,
          "textAlign": "center",
        }
      }
    >
      Feb 14
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#355974",
          "fontSize": 20,
          "textAlign": "center",
        }
      }
    >
      2017
    </Text>
  </Component>
  <Component
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    />
    <Component
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#818181",
            "fontSize": 14,
            "marginTop": 5,
          }
        }
      >
        a few seconds ago
      </Text>
      <Component
        style={
          Object {
            "backgroundColor": "#4A90E2",
            "borderRadius": 5,
            "marginLeft": 6,
            "marginTop": 5,
            "paddingHorizontal": 5,
            "paddingVertical": 1,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "white",
              "fontSize": 13,
            }
          }
        >
          every year
        </Text>
      </Component>
    </Component>
  </Component>
</View>
`;

exports[`Pages Reminders should renders correctly with reminders and fetching 3`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Reminders"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={
      Array [
        Object {
          "body": "My body",
          "created_at": "1991-12-17",
          "frequency_type": "month",
        },
        Object {
          "body": "My body 2",
          "created_at": "1993-12-17",
          "frequency_type": "all_time",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Component>
`;

exports[`Pages Reminders should renders correctly with reminders and fetching 4`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "flexDirection": "row",
      "paddingHorizontal": 15,
      "paddingVertical": 20,
    }
  }
>
  <Component
    style={
      Object {
        "paddingRight": 20,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#355974",
          "fontSize": 17,
          "textAlign": "center",
        }
      }
    >
      Feb 14
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#355974",
          "fontSize": 20,
          "textAlign": "center",
        }
      }
    >
      2017
    </Text>
  </Component>
  <Component
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    />
    <Component
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#818181",
            "fontSize": 14,
            "marginTop": 5,
          }
        }
      >
        a few seconds ago
      </Text>
      <Component
        style={
          Object {
            "backgroundColor": "#4A90E2",
            "borderRadius": 5,
            "marginLeft": 6,
            "marginTop": 5,
            "paddingHorizontal": 5,
            "paddingVertical": 1,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "white",
              "fontSize": 13,
            }
          }
        >
          every month
        </Text>
      </Component>
    </Component>
  </Component>
</View>
`;

exports[`Pages Reminders should renders correctly with reminders and fetching 5`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "flexDirection": "row",
      "paddingHorizontal": 15,
      "paddingVertical": 20,
    }
  }
>
  <Component
    style={
      Object {
        "paddingRight": 20,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#355974",
          "fontSize": 17,
          "textAlign": "center",
        }
      }
    >
      Feb 14
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#355974",
          "fontSize": 20,
          "textAlign": "center",
        }
      }
    >
      2017
    </Text>
  </Component>
  <Component
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    />
    <Component
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#818181",
            "fontSize": 14,
            "marginTop": 5,
          }
        }
      >
        a few seconds ago
      </Text>
      <Component
        style={
          Object {
            "backgroundColor": "#4A90E2",
            "borderRadius": 5,
            "marginLeft": 6,
            "marginTop": 5,
            "paddingHorizontal": 5,
            "paddingVertical": 1,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "white",
              "fontSize": 13,
            }
          }
        >
          frequency.all_time
        </Text>
      </Component>
    </Component>
  </Component>
</View>
`;

exports[`Pages Reminders should renders correctly without reminders and fetching 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Reminders"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Component>
`;

exports[`Pages Reminders should renders correctly without reminders and not fetching 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Reminders"
  />
  <EmptyActivity
    image={1}
    subtitle="Your memory might let you down - we won’t."
    title="Be reminded at the right time, for things that matter."
  />
</Component>
`;
